Feature/ Manage Booking Questions

hariombalhara opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes #2440 
Fixes #5536 
Fixes #5783 
Fixes #4915 
Fixes #5564 
Fixes #3350 
Fixes #5245
Fixes #6690
Implements https://github.com/calcom/cal.com/discussions/5840

## Looms
[Demo of sources and how Workflow adds smsReminderNumber now](https://www.loom.com/share/a59edc20ba874482ba86001f3416faf6)
[Introduction to Managing Booking Questions](https://www.loom.com/share/179f1d58940d4d03801c8dffca57f4e3)

## Short Term Follow-up
- #3350
- #5784
- [ ] Reuse Checkbox component cc @CarinaWolli 
- [ ] Support `responses` and `bookingFields` in API using feature detection(i.e. if `responses` are provided ignore `customInputs`). 
- [ ] Error Handling Improvements on Booking Page
	- Show Error Rings on fields
	- Validate on blur
- [x] Booking Success Page - User Fields to match with [Figma](https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr/Cal---New-Features?node-id=2519%3A135108&t=cLxDDR8ceL8T1DZz-1)
- [ ] `toString` implementation of each field to render responses differently for Email, Calendar, Booking Success Page. Currently, toString simply shows the response as a string. We have the context and we can render the responses with that context, e.g. Phone Number response should be a link that dials the phone number.
- [x] E2E Tests
	- Verify that the chosen location is actually working -> Verify the location shown on response page
	- Add a user field(one of each type) and check it is shown and can be filled on Booking Page and response is shown on booking success page
- [ ] Measure performance improvement of lazy loading FormBuilder on BookingPages( cc @roae Adding it to followup.)

## Long Term Follow-up
- Remove deprecated columns -` booking.customInputs`,` eventType.customInputs`, `eventType.disableGuests`, `eventType.metadata.additionalNotesRequired`
- **Tech Debt**: Make Routing Form use the same FormBuilder with a **variant** if needed.
- While Prefilling - `type`, `date`, `slug`, `count` are reserved query params for existing logic. More might be added. Adding a field with the same `name` would end up prefilling those with invalid information. Maybe restrict adding fields with those names or introduce a namespace for prefill query params?

**Note**: Migration introduces new columns only, so it's safe to be part of this PR itself. Instead of migrating customInputs to new schema for all existing rows, I am transforming from `customInputs` to `bookingFields` on the run, this allows us to avoid corrupting the data during migration.An eventType when saved, copies it's data to `bookingFields`, otherwise there is no modification in existing data. We can later plan to run a migration to do away with `customInputs` 

## Highlights
- Lesser number of Advanced Settings with the same configuration capability
   - **Disable Guests** - It isn't a separate setting now. It is controllable using the Booking Fields management
   - **Require Additional Notes** -  It isn't a separate setting now. It is controllable using the Booking Fields management 	
- **Enhancement**: Modifying Label for a question/field won't change the webhook payload or prefill query param. Earlier when you changed the label for a custom input, the payload received in Webhook and prefill query param name would change as they were generated from Label.
-  **BugFix**: There was a bug earlier that a Custom Input of Radio type wasn't prefilled. It is prefilled correctly now.
- The booking page would not bear the burden of pulling in fields from different places now. e.g. BookingPage was pulling  in data from workflow to determine if a field is to be shown. Now, workflow would push that field to bookingFields directly.

## Backward Compatibility
- Booking links with prefill params even prefill params for **customInputs** would keep on working . This is achieved by transforming `customInputs` into `bookingFields` using **slugified customInput label** as the **name** for the `bookingField`
- Any existing webhook endpoint would keep on receiving `customInputs` in the exact same format. This is achieved by transforming `booking.responses` into `customInputs` when a booking is done.
- Duplicating an existing EventType would still duplicate `customInputs`. This would allow the situation where an EventType which didn't have its `customInputs` saved into `bookingFields` would keep on working.
- A booking that was created before the EventType's `customInputs` were mapped to `bookingFields` would still be able to show responses by a mapping of `booking.customInputs` to `booking.responses`
- An EventType which is connected to an Attendee SMS workflow would add the `smsReminderNumber` field on the fly. It allows that EventType to still collect `smsReminderNumber` data without modifying the corresponding table row.

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] New feature (non-breaking change which adds functionality)

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- I haven't added tests that prove my fix is effective or that my feature works

hariombalhara committed:
WIP

hariombalhara committed:
Create Booking Questions builder

hariombalhara committed:
Renaming things

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
wip

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
wip

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Implement Add Guests and other fixes

hariombalhara committed:
Fixes after testing

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Fix wrong status code 404

hariombalhara committed:
Fixes

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Lint fixes

hariombalhara committed:
Self review comments addressed

markflorkowski commented:
Per @PeerRich's suggestion, asking here --

are user responses to booking questions available in metadata for use in custom webhooks?

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

zomars commented on apps/web/pages/booking/[uid].tsx:
```diff
@@ -574,15 +573,40 @@ export default function Success(props: SuccessProps) {
+                      // We show location in the "where" section
+                      // We show Booker Name, Emails and guests in Who section
+                      // We show notes in additional notes section
+                      if (
+                        !field ||
+                        field.name === "location" ||
+                        field.name === "guests" ||
+                        field.name === "notes" ||
+                        field.name === "name" ||
+                        field.name === "email"
+                      ) {
+                        return null;
+                      }
```
https://handbook.cal.com/engineering/best-practices/avoid-comparing-multiple-values-with-includes

```suggestion
                      if (!field || ["location", "guests", "notes", "name", "email"].includes(field.name)) {
                        return null;
                      }
```

zomars commented on apps/web/pages/event-types/[type]/index.tsx:
```diff
@@ -176,48 +177,59 @@ const EventTypePage = (props: EventTypeSetupProps) => {
     ),
   });
 
+  useEffect(() => {
+    if (!formMethods.formState.isDirty) {
+      //TODO: What's the best way to sync the form with backend
+      formMethods.setValue("bookingFields", defaultValues.bookingFields);
+    }
+  }, [defaultValues]);
+
```
Don't render the form until defaultValues is not undefined.

```suggestion
```

[Try the React Container approach](https://reactpatterns.js.org/docs/container-component)

zomars commented on packages/features/bookings/lib/getBookingResponsesSchema.ts:
```diff
@@ -0,0 +1,163 @@
+          return;
+        }
+
+        if (
+          bookingField.type === "address" ||
+          bookingField.type === "text" ||
+          bookingField.type === "select" ||
+          bookingField.type === "name" ||
+          bookingField.type === "number" ||
+          bookingField.type === "radio" ||
+          bookingField.type === "textarea"
+        ) {
```
https://handbook.cal.com/engineering/best-practices/avoid-comparing-multiple-values-with-includes

zomars commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -305,43 +312,145 @@ async function ensureAvailableUsers(
+  eventType,
+}: {
+  req: NextApiRequest;
+  isNotAnApiCall: boolean;
+  eventType: Awaited<ReturnType<typeof getEventTypesFromDB>>;
```
```suggestion
    args: { isNotAnApiCall: boolean;  eventType: Awaited<ReturnType<typeof getEventTypesFromDB>>; }
```

zomars replied:
I just don't want to pollute the function with extra arguments. I see these as ` args` or `options`

zomars commented on packages/lib/hooks/useHasPaidPlan.ts:
```diff
@@ -13,7 +13,7 @@ export function useHasPaidPlan() {
   const hasPaidPlan = hasTeamPlan?.hasTeamPlan || isCurrentUsernamePremium;
 
-  return { isLoading, hasPaidPlan };
+  return { isLoading, hasPaidPlan: hasPaidPlan };
```
```suggestion
  return { isLoading, hasPaidPlan };
```

zomars commented on packages/ui/components/form/inputs/Input.tsx:
```diff
@@ -338,6 +339,8 @@ const PlainForm = <T extends FieldValues>(props: FormProps<T>, ref: Ref<HTMLForm
           form
             .handleSubmit(handleSubmit)(event)
             .catch((err) => {
+              console.log(err);
```
```suggestion
```

zomars commented on packages/ui/components/form/inputs/Input.tsx:
```diff
@@ -338,6 +339,8 @@ const PlainForm = <T extends FieldValues>(props: FormProps<T>, ref: Ref<HTMLForm
             .handleSubmit(handleSubmit)(event)
             .catch((err) => {
+              console.log(err);
+              // FIXME: Booking Pages don't have toast, so this error is never shown
```
Can we dynamic load them? I don't want to increase booking pages bundle size any more.

zomars commented on packages/ui/form/PhoneInput.tsx:
```diff
@@ -1,29 +1,21 @@
-import BasePhoneInput, { Props } from "react-phone-number-input/react-hook-form";
+import BasePhoneInput, { Props } from "react-phone-number-input";
```
Can we lazy load this component?

hariombalhara replied:
Yup should be doable. Raised a separate issue.

zomars reviewed (changes requested):
So far so good. Got some feedback and need to test more as well.

hariombalhara commented:
@markflorkowski Yes, they are available right now as `customInputs` and would be available after this PR in `responses`. `customInputs` would still remain there for backward compatibility.

hariombalhara committed:
More self review comments addressed

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara committed:
Feedback from zomars

hariombalhara committed:
BugFixes after testing

hariombalhara committed:
More fixes discovered during review

hariombalhara committed:
Update packages/lib/hooks/useHasPaidPlan.ts

Co-authored-by: Omar López <zomars@me.com>

hariombalhara commented:
@CarinaWolli The approach to ask for smsReminderNumber has changed in this PR, requesting your review on that.

hariombalhara committed:
More fixes discovered during review

hariombalhara committed:
More fixes discovered during review

hariombalhara committed:
Update packages/ui/components/form/inputs/Input.tsx

Co-authored-by: Omar López <zomars@me.com>

hariombalhara committed:
Merge branch 'feat/manage-all-booking-inputs' of github.com:calcom/cal.com into feat/manage-all-booking-inputs

sean-brydon commented on packages/app-store/ee/routing-forms/components/react-awesome-query-builder/widgets.tsx:
```diff
@@ -25,49 +79,70 @@ const TextAreaWidget = (props: TextWidgetProps) => {
     setValue(val);
   };
   const textValue = value || "";
+  if (noLabel) {
+    return (
+      <Input
+        type={type}
+        className="dark:placeholder:text-darkgray-600 focus:border-brand dark:border-darkgray-300 dark:text-darkgray-900 block w-full rounded-md border-gray-300 text-sm focus:ring-black disabled:bg-gray-200 disabled:hover:cursor-not-allowed dark:bg-transparent dark:selection:bg-green-500 disabled:dark:text-gray-500"
+        value={textValue}
+        placeholder={placeholder}
+        disabled={readOnly}
+        onChange={onChange}
+        {...remainingProps}
+        {...customProps}
+      />
+    );
+  }
```
we could simply this by just adding 

`labelSrOnly={noLabel}` to the text field below. This will visually disable the label for the TextField but also still provide the input name to the htmlFor so we still support the a11y requirements here 

(If im not understanding whats happening here let me know :D )

hariombalhara replied:
@sean-brydon This is exactly what I need. Completely missed it. Thanks 🙏

sean-brydon commented on packages/features/bookings/lib/getBookingFields.ts:
```diff
@@ -0,0 +1,245 @@
+
+type Fields = z.infer<typeof eventTypeBookingFields>;
+
+const EventTypeCustomInputType = {
+  TEXT: "TEXT",
+  TEXTLONG: "TEXTLONG",
+  NUMBER: "NUMBER",
+  BOOL: "BOOL",
+  RADIO: "RADIO",
+  PHONE: "PHONE",
+};
```
I believe this would provide better TS support when using this 
```suggestion
const EventTypeCustomInputType = {
  TEXT: "TEXT",
  TEXTLONG: "TEXTLONG",
  NUMBER: "NUMBER",
  BOOL: "BOOL",
  RADIO: "RADIO",
  PHONE: "PHONE",
} as const;
```

sean-brydon commented on packages/features/bookings/lib/getBookingFields.ts:
```diff
@@ -0,0 +1,245 @@
+    },
+  ];
+
+  const missingSystemBeforeFields = [];
+  for (const field of systemBeforeFields) {
+    // Only do a push, we must not update existing system fields as user could have modified any property in it,
+    if (!bookingFields.find((f) => f.name === field.name)) {
+      missingSystemBeforeFields.push(field);
+    }
+  }
```
NIT: I believe this could be a map (would save milliseconds)  / look a little cleaner

sean-brydon replied:
Makes sense :)

sean-brydon commented on packages/features/form-builder/Components.tsx:
```diff
@@ -0,0 +1,379 @@
+// TODO: Share FormBuilder components across react-query-awesome-builder(for Routing Forms) widgets.
+// There are certain differences b/w two. Routing Forms expect label to be provided by the widget itself and FormBuilder adds label itself and expect no label to be added by component.
+// Routing Form approach is better as it provides more flexibility to show the label in complex components. But that can't be done right now because labels are missing consistent asterisk required support across different components
+export const Components: Record<BookingFieldType, Component> = {
```
This is awesome @hariombalhara 🔥

sean-brydon commented on packages/ui/components/form/inputs/Label.tsx:
```diff
@@ -4,7 +4,11 @@ export function Label(props: JSX.IntrinsicElements["label"]) {
     <label
       {...props}
-      className={classNames("mb-2 block text-sm font-medium leading-none text-gray-700", props.className)}>
+      // TODO: Test labels inside the app when dark theme is enabled as app doesn't support dark theme
+      className={classNames(
+        "mb-2 block text-sm font-medium leading-none text-gray-700 dark:text-white",
+        props.className
+      )}>
```
Youre all good here we have a constant

`isThemeSupported` which is only enabled on booking pages - else we only allow light mode :)

hariombalhara replied:
Thanks @sean. I tested as well. Removing the comment

hariombalhara committed:
Update packages/features/bookings/lib/getBookingFields.ts

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

hariombalhara committed:
More PR review fixes

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge branch 'feat/manage-all-booking-inputs' of github.com:calcom/cal.com into feat/manage-all-booking-inputs

CarinaWolli reviewed (dismissed):
Tested everything around SMS Reminder Number and I found the following issues: 
- Translation is missing for "SMS Reminder Number" (translation key is "number_sms_notifications")
- I think the input field used to be editable when rescheduling. I am also open to have it disabled, but then it should be greyed out to make it clear. 
- When the location of the event type is "Attendee Phone Number" the phone number input is not shown (has nothing to do with SMS Reminder Number)

hariombalhara committed:
Hide label using labelSrOnly

hariombalhara commented:
Thanks a lot @CarinaWolli, I fixed the issues you mentioned. Found that the placeholder was also not configured and fixed that too for SMS reminder.

@CarinaWolli Attendee Phone Input [seems to be working for me.]( https://www.loom.com/share/ee064d58a42943f3ac087c7158e79622) Can you share a loom of the issue if you can see it still.

Also, while testing it again, I found another case that I thought I had handled already(but I guess it didn't because I changed my approach midway). Earlier, phone number input was asked only once no matter it was for location(Attendee Phone input) or for `smsReminder`. Now, 2 separate inputs are being asked for.  I will see if I can handle it in the Form Builder.

hariombalhara committed:
Fix Carinas feedback and implement 2 workflows thingy

hariombalhara reviewed (commented):
~~Blocker: Discovered a missing backward compatibility case. smsReminderNumber is not adding a new field for existing event types~~
Edit: Resolved

CarinaWolli commented:
@hariombalhara I tried it again and it shows the location 'Attendee Phone Number' correctly, but when booking it doesn't ask me for my phone number and the location shows a cal video link. Here is a loom, let me know if you need more details 🤔

https://www.loom.com/share/c406ce98029d4669b2672617a9530b53

hariombalhara commented:
Oh got it. That's an edge case I didn't think of when there is just attendee phone no location and nothing else. 
Thankyou 🙏🙏🙏

hariombalhara committed:
Misc fixes

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

hariombalhara committed:
Fixes from Loom comments and PR

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara commented:
@CarinaWolli Fixed the Attendee Phone issue along with backward compatibility for `smsReminderNumber` field. All good now.

hariombalhara commented:
> @hariombalhara I tried it again and it shows the location 'Attendee Phone Number' correctly, but when booking it doesn't ask me for my phone number and the location shows a cal video link. Here is a loom, let me know if you need more details 🤔
> 
> [loom.com/share/c406ce98029d4669b2672617a9530b53](https://www.loom.com/share/c406ce98029d4669b2672617a9530b53)


@CarinaWolli Fixed the Attendee Phone issue along with backward compatibility for smsReminderNumber field. All good now.

hariombalhara reviewed (commented):
Self review done.

cc @zomars. All comments resolved as well. I would still want more people to test it out, it being the very critical part.

I am still doing more testing.

hariombalhara committed:
Fix a lint errr

hariombalhara committed:
Fix cancellation reason

CarinaWolli commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,734 @@
+              />
+              <DialogFooter>
+                <DialogClose color="secondary">Cancel</DialogClose>
+                <Button type="submit">{isFieldEditMode ? t("save") : t("add")}</Button>
```
I can't edit fields. When clicking save this error is thrown:
![Screenshot 2023-02-15 at 10 04 40](https://user-images.githubusercontent.com/30310907/219066670-3520a631-4b2c-4a0f-b4ff-10d6f7d96377.png)

hariombalhara replied:
Ahh..regression due to recent name conflict check. Fixed it.

CarinaWolli commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,734 @@
+                  }}
+                />
+              ) : null}
+              <Controller
```
When editing a field I can't click 'yes' or 'no', it's working when creating a new field

hariombalhara replied:
@CarinaWolli I think you must be editing a system field, that doesn't allow marking itself optional. User fields can be marked yes or not in edit mode.

CarinaWolli replied:
I see, it works with the fields that I created. We should make that more clear that this is disabled and can't be edited

hariombalhara replied:
Yeah, I am not sure how to do it. But it is a valid point. @Jaibles I think a better disabled state for yes No button should help I guess.

hariombalhara replied:
Agreed, I too saw the problem. I was thinking that all buttons should be there but they should be disabled.

CarinaWolli replied:
Yes I think I like that better 👍 I just starts looking pretty messy once you change the order and you have system fields and your fields mixed up

CarinaWolli commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,734 @@
+                      Edit
+                    </Button>
+                    {field.editable !== "system" && field.editable !== "system-but-optional" && (
+                      <Button
+                        className="!p-0"
+                        color="minimal"
+                        onClick={() => {
+                          removeField(index);
+                        }}
+                        StartIcon={FiTrash2}
+                      />
```
```suggestion
                      <Button
                        color="minimal"
                        variant="icon"
                        onClick={() => {
                          removeField(index);
                        }}
                        StartIcon={FiTrash2}
                      />
```

hariombalhara committed:
Fix regression in edit due to name conflict check

CarinaWolli commented:
I am getting this error when having a Select field, I also have options defined

![Screenshot 2023-02-15 at 10 41 09](https://user-images.githubusercontent.com/30310907/219076515-a984c795-5aee-45eb-8058-40544f3e1d86.png)

hariombalhara commented on packages/features/bookings/lib/getBookingFields.ts:
```diff
@@ -0,0 +1,333 @@
+      defaultLabel: "location",
+      type: "radioInput",
+      name: "location",
+      // Even though it should be required it is optional in production with backend choosing CalVideo as the fallback
```
@CarinaWolli Yes, that should happen already. It is handled here https://github.com/calcom/cal.com/pull/6560/files#diff-d360af68e64bd72af3fbe53bea6bdfbf7a70e0e1366670d6adfd972c1059f88bR151

CarinaWolli replied:
I meant in the event-type settings. So this one: 
![Screenshot 2023-02-15 at 10 46 39](https://user-images.githubusercontent.com/30310907/219078154-4812488b-2f8a-4797-a5a0-4c11dbd3abcd.png)

But it might be ok to still have it here, was just a thought

hariombalhara replied:
Oh, so it should remain I think. The plan is to move location select to the right sidebar where we show locations anyway. There we always show the location if there is just one

hariombalhara committed:
Update packages/features/form-builder/FormBuilder.tsx

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

hariombalhara committed:
Fix options not set when default value is used

hariombalhara commented:
> I am getting this error when having a Select field, I also have options defined
> 
> ![Screenshot 2023-02-15 at 10 41 09](https://user-images.githubusercontent.com/30310907/219076515-a984c795-5aee-45eb-8058-40544f3e1d86.png)

@CarinaWolli Fixed this one as well 🙏 It was only when user didn't make any modification to options

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

zomars commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -305,43 +312,151 @@ async function ensureAvailableUsers(
     hasHashedBookingLink,
     language,
     appsStatus: reqAppsStatus,
-    ...reqBody
```
Isn't this the request body? Why the rename?

hariombalhara replied:
Technically it is not reqBody now.
For an API request the body has `name` but for request from frontend the body has `responses.name` and this detail is normalized using `getBookingData` method so that `bookingData` has name in both the cases.

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

zomars committed:
Restoring reqBody to avoid uneeded conflicts with main

CarinaWolli commented:
Another small thing I found: The input fields for 'Attendee Address' and 'Attendee phone number' are required, however it is not maked with *
![Screenshot 2023-02-15 at 15 50 26](https://user-images.githubusercontent.com/30310907/219157750-b4f9b41a-94e6-4e94-b7b6-308daa4bac3e.png)

Other than that everything is working for me now 👍

zomars commented:
Let's make it clearer that name CANNOT be optional:

<img width="561" alt="image" src="https://user-images.githubusercontent.com/3504472/219163780-4868230c-cdd3-45a3-93c1-a78abdbf4a30.png">

zomars commented on apps/web/components/booking/pages/BookingPage.tsx:
```diff
@@ -77,22 +56,112 @@ import { HashLinkPageProps } from "../../../pages/d/[link]/book";
+        let readOnly =
+          (field.editable === "system" || field.editable === "system-but-optional") && !!rescheduleUid;
+        let noLabel = false;
+        let hidden = false;
```
Why not?

```suggestion
        let hidden = !!field.hidden;
```

zomars replied:
BTW this solved this https://github.com/calcom/cal.com/pull/6560#issuecomment-1432026165

hariombalhara replied:
Yeah it should be that way only 🙏 It's a recent regression while fixing another bug

zomars reviewed (changes requested):
Still got more tests to do but will submit my feedback so far 🙏🏽

zomars committed:
Type fix

zomars commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,737 @@
+  const fieldType = field.type;
+  const componentConfig = Components[fieldType];
+
+  const isValueOfPropsType = (val: unknown, propsType: typeof componentConfig.propsType) => {
+    if (propsType === "text") {
+      return typeof val === "string";
+    }
+    if (propsType === "boolean") {
+      return typeof val === "boolean";
+    }
+    if (propsType === "textList") {
+      return val instanceof Array && val.every((v) => typeof v === "string");
+    }
+    if (propsType === "select") {
+      return typeof val === "string";
+    }
+    if (propsType === "multiselect") {
+      return val instanceof Array && val.every((v) => typeof v === "string");
+    }
+    if (propsType === "objectiveWithInput") {
+      return typeof value === "object" && value !== null ? "value" in value : false;
+    }
+    throw new Error(`Unknown propsType ${propsType}`);
+  };
```
```suggestion
  const isValueOfPropsType = (val: unknown, propsType: typeof componentConfig.propsType) => {
    const propsTypeConditionMap = {
      boolean: typeof val === "boolean",
      multiselect: val instanceof Array && val.every((v) => typeof v === "string"),
      objectiveWithInput: typeof value === "object" && value !== null ? "value" in value : false,
      select: typeof val === "string",
      text: typeof val === "string",
      textList: val instanceof Array && val.every((v) => typeof v === "string"),
    } as const;
    if (!propsTypeConditionMap[propsType]) throw new Error(`Unknown propsType ${propsType}`);
    return propsTypeConditionMap[propsType];
  };
```

zomars commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,737 @@
+  if (componentConfig.propsType === "multiselect") {
+    if (!field.options) {
+      throw new Error("Field options is not defined");
+    }
+    return (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          placeholder={field.placeholder}
+          readOnly={readOnly}
+          value={value as string[]}
+          setValue={setValue as (arg: typeof value) => void}
+          options={field.options.map((o) => ({ ...o, title: o.label }))}
+        />
+      </WithLabel>
+    );
+  }
+
+  if (componentConfig.propsType === "objectiveWithInput") {
+    if (!field.options) {
+      throw new Error("Field options is not defined");
+    }
+    if (!field.optionsInputs) {
+      throw new Error("Field optionsInputs is not defined");
+    }
+    return field.options.length ? (
+      <WithLabel field={field} readOnly={readOnly}>
+        <componentConfig.factory
+          placeholder={field.placeholder}
+          readOnly={readOnly}
+          name={field.name}
+          value={value as { value: string; optionValue: string }}
+          setValue={setValue as (arg: typeof value) => void}
+          optionsInputs={field.optionsInputs}
+          options={field.options}
+        />
+      </WithLabel>
+    ) : null;
+  }
+
+  throw new Error(`Field ${field.name} does not have a valid propsType`);
```
I'm pretty sure we can improve a lot in terms of legibility and typing here.

hariombalhara replied:
Yeah I think so too. But couldn't get type narrowing to work and had to resort to type assertion. Will followup on it.

zomars commented on packages/prisma/zod-utils.ts:
```diff
@@ -119,23 +141,26 @@ export const stringOrNumber = z.union([
 export const stringToDayjs = z.string().transform((val) => dayjs(val));
 
-export const bookingCreateBodySchema = z.object({
+export const bookingCreateSchemaForApiOnly = z.object({
```
Let's keep `bookingCreateBodySchema` intact and add a new one so we don't break API until it is update.

hariombalhara replied:
This is not breaking API. It has the old props for backward compatibility only. This is merged with `extendedBookingCreateBody` to get final schema. Maybe the name is a bit misleading, I will change that.

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,737 @@
+      label: "Select",
+      value: "select",
+      needsOptions: true,
+      isTextType: true,
+    },
+    {
+      label: "MultiSelect",
+      value: "multiselect",
+      needsOptions: true,
+      isTextType: false,
+    },
+    {
+      label: "Multiple Emails",
+      value: "multiemail",
+      isTextType: true,
+    },
+    {
+      label: "Radio Input",
+      value: "radioInput",
+      isTextType: false,
+      systemOnly: true,
+    },
+    {
+      label: "Checkbox Group",
+      value: "checkbox",
+      needsOptions: true,
+      isTextType: false,
+    },
+    {
+      label: "Radio Group",
+      value: "radio",
+      needsOptions: true,
+      isTextType: false,
+    },
+    {
+      label: "Checkbox",
+      value: "boolean",
+      isTextType: false,
+    },
+  ];
```
I am not sure if we can control the order of these types reliably now(across browsers).

hariombalhara replied:
Okay, I just read that the order is reliable since many years now. So, if property names are strings, they would be iterated in the order they are created. So this should be fine now.

hariombalhara commented on apps/web/components/booking/pages/BookingPage.tsx:
```diff
@@ -77,22 +56,112 @@ import { HashLinkPageProps } from "../../../pages/d/[link]/book";
+  rescheduleUid,
+  isDynamicGroupBooking,
+}: {
+  fields: BookingPageProps["eventType"]["bookingFields"];
```
I have a counter reasoning here. I am just passing on a prop created by getSSRProps here. If I end up generating a different variation of FieldsType in getSSRProps, the typecheck will pass, but it might crash at runtime.

zomars replied:
That's a problem of type signature differing from the runtime signature. It doesn't matter where this comes from as long as signature matches then there shouldn't be any problems.

hariombalhara committed:
Update apps/web/components/booking/pages/BookingPage.tsx

Co-authored-by: Omar López <zomars@me.com>

hariombalhara committed:
Update packages/features/form-builder/FormBuilder.tsx

Co-authored-by: Omar López <zomars@me.com>

hariombalhara committed:
Update apps/web/components/booking/pages/BookingPage.tsx

Co-authored-by: Omar López <zomars@me.com>

hariombalhara commented:
> Another small thing I found: The input fields for 'Attendee Address' and 'Attendee phone number' are required, however it is not maked with * ![Screenshot 2023-02-15 at 15 50 26](https://user-images.githubusercontent.com/30310907/219157750-b4f9b41a-94e6-4e94-b7b6-308daa4bac3e.png)
> 
> Other than that everything is working for me now 👍

Yeah that's known to me. I need to do some non trivial changes to get that working because that's not the field label you are seeing, that is actually the option label being shown as field label(for radioInput having just one option)

hariombalhara committed:
Apply suggestions from code review

Co-authored-by: Omar López <zomars@me.com>

hariombalhara committed:
Show fields but mark them disabled

hariombalhara commented:
> Let's make it clearer that name CANNOT be optional:

As discussed [here ](https://threads.com/34446204540?s=gno6fDuRTrF2SA1UmtqREK) we should be showing an asterisk consistently across the labels for all fields. That would solve the problem. I would pick it up as a followup

hariombalhara commented on apps/web/pages/booking/[uid].tsx:
```diff
@@ -537,63 +533,39 @@ export default function Success(props: SuccessProps) {
+                      // We show Booker Name, Emails and guests in Who section
+                      // We show notes in additional notes section
+                      // We show rescheduleReason at the top
+                      if (
+                        !field ||
+                        (
+                          [
+                            SystemField.Enum.location,
+                            SystemField.Enum.name,
+                            SystemField.Enum.email,
+                            SystemField.Enum.guests,
+                            SystemField.Enum.notes,
+                            SystemField.Enum.rescheduleReason,
+                          ] as string[]
+                        ).includes(field.name)
+                      ) {
+                        return null;
+                      }
```
Wow. That safeParse approach is much better. I unnecessarily made it complex. 
Thanks @zomars 🙏

hariombalhara committed:
Apply suggestions from code review

Co-authored-by: Omar López <zomars@me.com>

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -305,43 +312,151 @@ async function ensureAvailableUsers(
   const paymentAppData = getPaymentAppData(eventType);
 
-  // Check if required custom inputs exist
-  handleCustomInputs(eventType.customInputs as EventTypeCustomInput[], reqBody.customInputs);
```
It's needed for API request and moved to `getBookingData` fn.

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1246,7 +1359,7 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
     await scheduleWorkflowReminders(
       eventType.workflows,
-      reqBody.smsReminderNumber as string | null,
+      smsReminderNumber as string | null,
```
Turns out smsReminderNumber is not expected to be undefined by `scheduleWorkflowReminders` but it can be undefined. Handled that and removed this type assertion

hariombalhara committed:
More comments

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -712,12 +832,6 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
     return booking;
   }
 
-  if (reqBody.customInputs.length > 0) {
```
Moved to `getCustomInputsResponses` which handles backward compatibility as well.

hariombalhara commented:
Found another issue while testing **and it's fixed now.**
Steps:
1. You add no user fields and let the booker do a booking
2. Reach the booking/[uid] (success) page.
3. Add a user input and mark it required.
4. Refresh the already open booking/[uid] page. It would crash as it would expect that user field's response to be present but it will not be there because it's a booking that happened before the user field was added.

To fix this, I have created a partial version of bookingResponsesSchema. The non partial version should be used only when fresh booking is being created, otherwise we should use partial schema only.

cc @zomars @CarinaWolli for your review

hariombalhara committed:
Fix booking success page crash when a booking doesnt have newly added required fields response

hariombalhara committed:
Dark theme asterisk not visible

hariombalhara committed:
Make location required in zodSchema as was there in production

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

# Conflicts:
#	apps/web/public/static/locales/en/common.json

zomars commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -0,0 +1,736 @@
+                field={field}
+                value={value}
+                readOnly={readOnly}
+                setValue={(val: unknown) => {
+                  onChange(val);
+                }}
```
Why not let ComponentForField accept this signature?

```suggestion
                onChange={onChange}
```

zomars committed:
Linting

roae committed:
Merge branch 'main' into feat/manage-all-booking-inputs

# Conflicts:
#	apps/web/components/booking/pages/BookingPage.tsx
#	apps/web/components/eventtype/CustomInputTypeForm.tsx
#	apps/web/lib/getBooking.tsx
#	apps/web/lib/mutations/bookings/create-booking.ts
#	apps/web/lib/mutations/bookings/create-recurring-booking.ts
#	apps/web/pages/booking/[uid].tsx
#	apps/web/pages/event-types/[type]/index.tsx
#	apps/web/pages/team/[slug]/[type].tsx
#	apps/web/pages/team/[slug]/book.tsx
#	apps/web/public/static/locales/en/common.json
#	packages/lib/defaultEvents.ts
#	packages/lib/getEventTypeById.ts
#	packages/ui/form/AddressInput.tsx
#	packages/ui/form/PhoneInput.tsx

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Remove _metadata.ts files for apps that have config.json

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara committed:
Revert "Remove _metadata.ts files for apps that have config.json"

This reverts commit d79bdd336cf312a30a8943af94c059947bd91ccd.

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

roae committed:
Merge branch 'main' into feat/manage-all-booking-inputs

# Conflicts:
#	packages/features/bookings/lib/handleNewBooking.ts
#	packages/trpc/server/routers/viewer/workflows.tsx

roae committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

# Conflicts:
#	apps/web/components/booking/pages/BookingPage.tsx
#	apps/web/components/eventtype/CustomInputTypeForm.tsx
#	apps/web/lib/getBooking.tsx
#	apps/web/lib/mutations/bookings/create-recurring-booking.ts
#	apps/web/pages/booking/[uid].tsx
#	apps/web/pages/event-types/[type]/index.tsx
#	apps/web/pages/team/[slug]/[type].tsx
#	apps/web/public/static/locales/en/common.json
#	packages/features/bookings/lib/handleNewBooking.ts
#	packages/ui/form/AddressInput.tsx
#	packages/ui/form/PhoneInput.tsx

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara committed:
Fix lint error

hariombalhara committed:
Fix missing condition for samlSPConfig

hariombalhara committed:
Delete unexpectedly added file

hariombalhara committed:
yarn.lock change not required

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

roae committed:
fix types

roae committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

roae commented on apps/web/pages/team/[slug]/book.tsx:
```diff
@@ -2,9 +2,10 @@ import type { GetServerSidePropsContext } from "next";
 
 import type { LocationObject } from "@calcom/app-store/locations";
 import { privacyFilteredLocations } from "@calcom/app-store/locations";
+import { getBookingFieldsWithSystemFields } from "@calcom/features/bookings/lib/getBookingFields";
```
In this page we can also load the dependencies that are only needed in getServerSideProps within the function, we must take care of the bundle size of the public pages.

CarinaWolli commented on packages/features/form-builder/Components.tsx:
```diff
@@ -0,0 +1,385 @@
+          {options.map((option, i) => {
+            return (
+              <label key={i} className="block">
+                <input
```
can we use our `<Checkbox/>` component here? Our checkboxes should have rounded corners

hariombalhara replied:
Tried using <Checkbox> and it mandates a description field right now, which this component doesn't need because it is going to show a list of checkboxes. Need to think about making changes in the component. 
I am adding it to followup list in the PR and will pick it up. Don't want to do changes to more files in this PR, as it's already huge.

CarinaWolli replied:
Sounds good, maybe you can make rounded corners to the checkboxes for now. That's why I saw that we don't reuse the component here :)

hariombalhara replied:
Made it rounded.

roae reviewed (dismissed):
There are some opportunities to decrease the bundle size of some public pages, and I think after that we can work on reducing the amount of tailwind alvitriary values.

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara commented on apps/web/pages/team/[slug]/[type].tsx:
```diff
@@ -2,6 +2,7 @@ import type { GetServerSidePropsContext } from "next";
 
 import type { LocationObject } from "@calcom/core/location";
 import { privacyFilteredLocations } from "@calcom/core/location";
+import { getBookingFieldsWithSystemFields } from "@calcom/features/bookings/lib/getBookingFields";
```
I had thought that Next.js should automatically remove this through treeshaking. Because it knows it's used in getServerSideProps only.

Found an article on it https://gedasfx.medium.com/how-does-next-js-10-handle-imports-c51b9e4451c6#:~:text=All%20imports%2C%20which%20are%20then%20used%20ONLY,will%20be%20sent%20over%20to%20the%20client.

Also, I spent some time verifying that it is being removed automatically, but because of `getBookingFieldsWithSystemFields` being imported elsewhere(some indirect dependency), it's still being imported.

I will do followup work on some approach to remove it. 
I also checked Next.js bundle analyzer stats and the payload isn't changing much from what we have already.

hariombalhara replied:
Btw, while testing it out I found something else that we fix too, filed an issue for that [here](https://github.com/calcom/cal.com/issues/7276)

roae replied:
Yes, I had already read that post, but in booking page it reduces -100kb when I tried async imports, maybe the reduction was for another reason.

roae commented on apps/web/components/booking/pages/BookingPage.tsx:
```diff
@@ -648,392 +574,23 @@ const BookingPage = ({
-                )}
-
-                <div className="flex justify-end space-x-2 rtl:space-x-reverse">
+              <Form form={bookingForm} noValidate handleSubmit={bookEvent}>
```
Ok, you are right, loading it dynamically might slow down the LCP.

JeroenReumkens commented on apps/web/components/booking/pages/BookingPage.tsx:
```diff
@@ -1,34 +1,35 @@
 import { zodResolver } from "@hookform/resolvers/zod";
```
This will be fun solving conflicts with the new booker component  😅 That new component merges the booker + booking form into one page. 

Not much we you can do for now, but it's good to be aware.

See this draft PR from this file down for more context: https://github.com/calcom/cal.com/pull/6792/files#diff-3314f3429ad01dbd0242361878990240275a062a8e0d564b2ec04a856ffae24b

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Make checkboxes rounded

hariombalhara committed:
Fix defaultLabel being stored as label due to SSR rendering

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

zomars committed:
Shaved 16kb from booking page

zomars commented:
Shaved 16kb by lazy loading the payment data (not all event types use this)

<img width="841" alt="image" src="https://user-images.githubusercontent.com/3504472/221270847-67696d6e-f38e-4338-86d6-48e2e2a22a32.png">

<img width="843" alt="image" src="https://user-images.githubusercontent.com/3504472/221271115-6fcae8f8-80c1-49cc-bb74-550e1454ad59.png">

zomars committed:
Explicit types for profile

zomars committed:
Merge branch 'main' into feat/manage-all-booking-inputs

zomars reviewed (approved):
I think we're on good position to merge. Let's test this troughly on staging.

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/manage-all-booking-inputs' into feat/manage-all-booking-inputs

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Show payment value only if price is greater than 0

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara commented on apps/web/components/booking/BookingDescriptionPayment.tsx:
```diff
@@ -5,8 +5,7 @@ import { FiCreditCard } from "@calcom/ui/components/icon";
   const paymentAppData = getPaymentAppData(props.eventType);
-
-  if (!paymentAppData || paymentAppData.price > 0) return null;
+  if (!paymentAppData || paymentAppData.price <= 0) return null;
```
@zomars FYI, have changed this check as it was rendering only when payment isn't required..

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

hariombalhara committed:
Fix type error

hariombalhara committed:
Add back inferred types as they are failing

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/manage-all-booking-inputs

CarinaWolli commented:
The number field has 2 labels
<img width="362" alt="Screenshot 2023-03-02 at 09 48 23" src="https://user-images.githubusercontent.com/30310907/222462290-90aed830-1fb7-4202-98ea-fdcf9ce91acd.png">

hariombalhara commented:
> The number field has 2 labels <img alt="Screenshot 2023-03-02 at 09 48 23" width="362" src="https://user-images.githubusercontent.com/30310907/222462290-90aed830-1fb7-4202-98ea-fdcf9ce91acd.png">

What?? Why ?? :(. Let me fix it.

hariombalhara committed:
Fix duplicate label on number

CarinaWolli reviewed (approved):
Amazing work @hariombalhara 🙌🏻
